src/ledger-bridge-keyring.ts

```diff
@@ -0,0 +1,883 @@
+import * as ethUtil from 'ethereumjs-util';
+import * as sigUtil from 'eth-sig-util';
+import { TransactionFactory, TxData, TypedTransaction } from '@ethereumjs/tx';
+import type OldEthJsTransaction from 'ethereumjs-tx';
+import type LedgerHwAppEth from '@ledgerhq/hw-app-eth';
+
+const pathBase = 'm';
+const hdPathString = `${pathBase}/44'/60'/0'`;
+const keyringType = 'Ledger Hardware';
+
+const BRIDGE_URL = 'https://metamask.github.io/eth-ledger-bridge-keyring';
+
+const MAX_INDEX = 1000;
+const NETWORK_API_URLS: Record<string, string> = {
+  ropsten: 'http://api-ropsten.etherscan.io',
+  kovan: 'http://api-kovan.etherscan.io',
+  rinkeby: 'https://api-rinkeby.etherscan.io',
+  mainnet: 'https://api.etherscan.io',
+} as const;
+
+enum IFrameMessageAction {
+  LedgerConnectionChange = 'ledger-connection-change',
+  LedgerUnlock = 'ledger-unlock',
+  LedgerMakeApp = 'ledger-make-app',
+  LedgerUpdateTransport = 'ledger-update-transport',
+  LedgerSignTransaction = 'ledger-sign-transaction',
+  LedgerSignPersonalMessage = 'ledger-sign-personal-message',
+  LedgerSignTypedData = 'ledger-sign-typed-data',
+}
+
+interface IFrameMessage {
+  action: IFrameMessageAction;
+  params: Record<string, unknown>;
+  target: string;
+  messageId: number;
+}
+
+type GetAddressPayload = Awaited<ReturnType<LedgerHwAppEth['getAddress']>>;
+
+type SignMessagePayload = Awaited<
+  ReturnType<LedgerHwAppEth['signEIP712HashedMessage']>
+>;
+
+type SignTransactionPayload = Awaited<
+  ReturnType<LedgerHwAppEth['signTransaction']>
+>;
+
+interface ConnectionChangedPayload {
+  connected: boolean;
+}
+
+type IFrameMessageResponsePayload = { error?: unknown } & (
+  | GetAddressPayload
+  | SignTransactionPayload
+  | SignMessagePayload
+  | ConnectionChangedPayload
+);
+
+export interface IFrameMessageResponse {
+  success: boolean;
+  action: IFrameMessageAction;
+  messageId: number;
+  payload: IFrameMessageResponsePayload;
+  error?: unknown;
+}
+
+export interface AccountDetails {
+  index?: number;
+  bip44?: boolean;
+  hdPath?: string;
+}
+
+export interface LedgerBridgeKeyringOptions {
+  hdPath: string;
+  accounts: readonly string[];
+  accountDetails: Readonly<Record<string, AccountDetails>>;
+  accountIndexes: Readonly<Record<string, number>>;
+  bridgeUrl: string;
+  implementFullBIP44: boolean;
+}
+
+/**
+ * Check if the given transaction is made with ethereumjs-tx or @ethereumjs/tx
+ *
+ * Transactions built with older versions of ethereumjs-tx have a
+ * getChainId method that newer versions do not.
+ * Older versions are mutable
+ * while newer versions default to being immutable.
+ * Expected shape and type
+ * of data for v, r and s differ (Buffer (old) vs BN (new)).
+ *
+ * @param tx
+ * @returns Returns `true` if tx is an old-style ethereumjs-tx transaction.
+ */
+function isOldStyleEthereumjsTx(
+  tx: TypedTransaction | OldEthJsTransaction,
+): tx is OldEthJsTransaction {
+  return typeof (tx as OldEthJsTransaction).getChainId === 'function';
```

Comment:
I applied this suggestion to all the functions used for narrowing types, but not on this one.
For some reason `hasProperty(tx, 'getChainId')` returns false also in case of an old style transaction. I suspect this happens because EthereumTx is a function and `getChainId` is added to its prototype when instantiating.

If you have other solutions to make this work I'll be happy to commit them, for now I left this one as-is.

Reply:
Oh! Right, that's because `hasProperty` just checks own properties. But since `tx` is an instance, we need to check for the property in its prototype. So maybe this would be more appropriate?

``` suggestion
  return 'getChainId' in tx && typeof tx.getChainId === 'function';
```